### ISSUE TITLE
Warn on push_str(&format!())

### ISSUE THREAD

**Issue by kangalioo** [rust-lang/rust-clippy#6261]
### What it does

Emits a warning on code like ```rust
some_string.push_str(&format!(...));
// or
some_string += &format!(...);
```
and suggests rewriting it to ```rust
write!(&mut some_string, ...);
```

### Categories
- Kind: `Perf`

*What is the advantage of the recommended code over the original code*
It avoids one extra heap allocation by writing directly to the existing String instead of creating a new temporary String.

### Drawbacks
The user has to `use std::io::Write` for the `write!` macro to work (which I don't get tbh, why can't `write!` just import the trait on its own...)

### Example
```rust
let mut string = String::new();
for foo in foo_list {
    string += &format!("{:?}", foo);
}
```

Could be written as:

```rust
use std::io::Write;

let mut string = String::new();
for foo in foo_list {
    write!(&mut string, "{:?}", foo);
}
```

**Comment by Urcra** [comment]
I had actually thought this would be optimized away but seems like the slowdown is quite significant in release mode see example here https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=daed734101a8d8bc0c6a7bd811154504 

There are some additional drawbacks in addition to the ones you mentioned however. `write!` returns a result so you would need to write it as:

```rust
use std::fmt::Write;

let mut string = String::new();
for foo in foo_list {
    write!(&mut string, "{:?}", foo).expect("This can never fail");
}
```

Which might cause other lints to give you trouble specifically lints such as `unwrap_in_result` if using this piece of code in a function that returns a `Result`

You also need to import either/both `std::fmt:Write` and `std::io::Write` and in the case of importing both of them you need to give them an alias so they don't clash such as:

```rust
use std::fmt::Write as FmtWrite;
use std::io::Write as IoWrite;
```

I also feel the readability of `string += &format!("{:?}", foo);` is better than `write!(&mut string, "{:?}", foo).expect("This can never fail");` Though that is more of a subjective opinion

All that said I still feel like the performance probably makes up for the drawbacks.

I'll try to implement a lint for this over weekend

**Comment by camsteffen** [comment]
If anyone asks about using `write!` with `String`, just change the subject to how amazing Rust's ownership system is or something like that.
